tmk_core/common/chibios/eeprom_stm32.c

```diff
@@ -48,72 +58,124 @@ uint16_t EEPROM_Init(void) {
  *  Erase the whole reserved Flash Space used for user Data
  ******************************************************************************/
 void EEPROM_Erase(void) {
-    int page_num = 0;
+    // erase all flash pages
+    EEPROM_Clear();
 
-    // delete all pages from specified start page to the last page
-    do {
-        FLASH_ErasePage(FEE_PAGE_BASE_ADDRESS + (page_num * FEE_PAGE_SIZE));
-        page_num++;
-    } while (page_num < FEE_DENSITY_PAGES);
+    // reset the content of the buffer
+    memset(&DataBuf[0], FEE_EMPTY_BYTE, sizeof(DataBuf));
 }
+
 /*****************************************************************************
- *  Writes once data byte to flash on specified address. If a byte is already
- *  written, the whole page must be copied to a buffer, the byte changed and
- *  the manipulated buffer written after PageErase.
+ * Write data with its eeprom address to flash if there has empty words,
+ * otherwise backup the current valid data, erase all flash pages,
+ * and finally restore the valid data.
  *******************************************************************************/
 uint16_t EEPROM_WriteDataByte(uint16_t Address, uint8_t DataByte) {
     FLASH_Status FlashStatus = FLASH_COMPLETE;
 
-    uint32_t page;
-    int      i;
-
-    // exit if desired address is above the limit (e.G. under 2048 Bytes for 4 pages)
-    if (Address > FEE_DENSITY_BYTES) {
+    uint32_t addr;
+    // exit if not a valid address
+    if (!IS_VALID_ADDRESS(Address)) {
         return 0;
     }
 
-    // calculate which page is affected (Pagenum1/Pagenum2...PagenumN)
-    page = FEE_ADDR_OFFSET(Address) / FEE_PAGE_SIZE;
+// we are sure the address will not be out of bound
+#pragma GCC diagnostic ignored "-Warray-bounds"
+    // same value, do not need program
+    if (DataBuf[Address] == DataByte) {
+        return FlashStatus;
+    }
 
-    // if current data is 0xFF, the byte is empty, just overwrite with the new one
-    if ((*(__IO uint16_t *)(FEE_PAGE_BASE_ADDRESS + FEE_ADDR_OFFSET(Address))) == FEE_EMPTY_WORD) {
-        FlashStatus = FLASH_ProgramHalfWord(FEE_PAGE_BASE_ADDRESS + FEE_ADDR_OFFSET(Address), (uint16_t)(0x00FF & DataByte));
+    // find the address can be written
+    addr = EEPROM_FindValidAddress();
+    DataBuf[Address] = DataByte;
+#pragma GCC diagnostic pop
+    if (addr == FEE_EMPTY_VALUE) {
+        // EEPROM is full, need to erase and reprogramming
```

Comment:
there's an issue here that when your eeprom is completely full, every single byte write will perform a full erase. I think it would be good to add some "slack" e.g. reserve twice the flash memory compared to emulated eeprom size (so times 8 instead of current times 4)